openedx/core/djangoapps/safe_sessions/tests/test_middleware.py

```diff
@@ -615,3 +622,544 @@ def test_user_change_with_no_ids(self):
+        """
+        Calls EmailChangeMiddleware.process_response when user is authenticated and email
+        change was requested.
+        Verify that the logged-in cookies are updated
+        """
+        # return value of mock
+        mock_set_logged_in_cookies.return_value = self.client.response
+
+        # Log in the user
+        self.client.login(email=self.user.email, password=self.PASSWORD)
+        self.request.session = self.client.session
+        self.client.response.set_cookie(settings.SESSION_COOKIE_NAME, 'authenticated')  # Add some logged-in cookie
+
+        # Registering email change (setting a variable `email_change_requested` to indicate email was changed)
+        # so that process_response can update cookies
+        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
+
+        # Call process_response
+        response = EmailChangeMiddleware(get_response=lambda request: None).process_response(
+            self.request, self.client.response
+        )
+
+        assert response.status_code == 200
+        # Assert that cookies are updated
+        # Assert that mock_set_logged_in_cookies is called
+        mock_set_logged_in_cookies.assert_called()
+
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_logged_in_cookies")
+    def test_process_response_email_change_not_requested(self, mock_set_logged_in_cookies):
+        """
+        Calls EmailChangeMiddleware.process_response when user is authenticated but email
+        change was not requested.
+        Verify that the logged-in cookies are not updated
+        """
+        # return value of mock
+        mock_set_logged_in_cookies.return_value = self.client.response
+
+        # Log in the user
+        self.client.login(email=self.user.email, password=self.PASSWORD)
+        self.request.session = self.client.session
+        self.client.response.set_cookie(settings.SESSION_COOKIE_NAME, 'authenticated')  # Add some logged-in cookie
+
+        # No call to register_email_change to indicate email was not changed
+
+        # Call process_response
+        response = EmailChangeMiddleware(get_response=lambda request: None).process_response(
+            self.request, self.client.response
+        )
+
+        assert response.status_code == 200
+        # Assert that cookies are not updated
+        # Assert that mock_set_logged_in_cookies not called
+        mock_set_logged_in_cookies.assert_not_called()
+
+    @patch.dict("django.conf.settings.FEATURES", {"DISABLE_SET_JWT_COOKIES_FOR_TESTS": False})
+    def test_user_remain_authenticated_on_email_change_in_other_browser_with_toggle_disabled(self):
+        """
+        Integration Test: test that a user remains authenticated upon email change
+        in other browser when ENFORCE_SESSION_EMAIL_MATCH toggle is disabled
+        Verify that the session and cookies are not affected
```

Comment:
> I thought the cookies will be updated after email change regardless of the toggle setting? Am I incorrect, or should this also be verified?

Yes, you are right. The cookies get updated in the browser where the email change happened. Not in other browsers where the user may have sessions. This test verifies that when the toggle is **disabled** and email is changed in some other browser, the cookies of current browser are not affected (not cleared) and the user remains authenticated in the current browser. On the other case, if the toggle was enabled and email was changed in some other browser, the user would have been logged out of current browser. 

Although its true that the cookies of browser in which the email change has happened, will be updated regardless of toggle setting.

I think I may need to update the description of test for better clarification.

Reply:
> I'm not sure how complicated it would be to instantiate a second Django client to represent the other browser, and actually make requests, rather than dropping back to mocking again. Feel free not to not pursue this, especially given the time invested. However, if pursued, maybe a longer test could show the implications in both clients with some test combining.

For now, we won't be working on this one considering the time invested instead we have created a ticket in our backlog to work on this improvement in tests: https://2u-internal.atlassian.net/browse/VAN-1799. We will surely pick it up and work on it.

Reply:
I think you should forget about this. Tests are good. I added a comment to the ticket. I'm going to mark this resolved as well.